### DISCUSSION TITLE
How to get app port?

### DISCUSSION THREAD

**Discussion by BaselOmari** [fastapi/fastapi#9703] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()

@app.on_event("startup")
async def startup_event():
   port = app.port
   print("The port used for this app is", port)
```

### Description

Assuming I don't specify a port number, the application gets run on an available port (`Uvicorn running on http://127.0.0.1:8000`). Is there a way to extract what the port selected is and use it as a constant in the app?

### Operating System

Linux

### FastAPI Version

0.95.1

### Python Version

Python 3.10.11

### Additional Context

For more context, when starting up an app, I wish to get the port number used, and send it to another service, such that it can direct any requests to this address.

**Comment by jgould22** [comment]
Couple of things.

> Assuming I don't specify a port number, the application gets run on an available port

I don't think its an available port, 8000 is just Uvicorns default, if something is using that port then it will probably throw an error.

> For more context, when starting up an app, I wish to get the port number used, and send it to another service, such that it can direct any requests to this address.

So the ASGI Application (in this case fastapi) doesn't really concern itself with what port it is getting data from. That is more on the level of the ASGI Server (Uvicorn in this case).

You have a few options. 

If you are using Gunicorn (instead of uvicorn) there are server hooks that you can use to send the data with (https://docs.gunicorn.org/en/stable/settings.html#server-hooks) 

If you want to use uvicorn an option might be to load your settings before running uvicorn programmatically at which point you can send your port data (since you will set it in the settings)

**Reply by BaselOmari** [reply]
I assumed that uvicorn finds the first available port (i.e. if 8000 wasn't available, it would use 8001). To get that specific feature, I'll be setting `port=0` when using `uvicorn.run`.

I'm not looking to have any hardcoded port values, as I'll be running tens of these clients using the same script (`python client.py`) and can't pass unique parameters to each.

I've found [this discussion thread](https://github.com/encode/uvicorn/issues/761) which highlights a solution at the end. I might be combining that with the server hooks that you kindly mentioned.

Thanks for your contribution!

**Reply by jgould22** [reply]
I did not know you could set the port=0, cool.

**Comment by abhint** [comment]
```py 
import uvicorn
from fastapi import FastAPI

app = FastAPI()

PORT: int  = 5000


@app.on_event("startup")
def startup_event():
    print("The port used for this app is", PORT)

@app.get("/")
def root():
    return {"message": "Hello, World!"}

if __name__ == "__main__":
    uvicorn.run(app, host="127.0.0.1", port=PORT)  
```

**Reply by BaselOmari** [reply]
I'm not looking to have any hardcoded port values, as I'll be running tens of these clients using the same script (python client.py) and can't pass unique parameters to each.

Thanks for your contribution!

**Comment by StEvUgnIn** [comment]
```python
from fastapi import FastAPI
from uvicorn import Config

app = FastAPI()

@app.on_event("startup")
async def startup_event():
   port = Config(app).port
   print("The port used for this app is", port)
```
